### PR TITLE
Convert additional database methods to async (select list, search, insert_many, delete_*)

### PR DESCRIPTION
**Pull request by clokep**
This converts `simple_select_list_paginate`, `simple_search_list`, `simple_insert_many`, `execute`, `simple_delete_one`, and `simple_delete_many` to async/await.

It also converts calls to async/await and fixes some types hints / comments.

This is essentially #8162, but different database methods.

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -1512,7 +1529,7 @@ def get_cache_dict(
 
         return cache, min_val
 
-    def simple_select_list_paginate(
+    async def simple_select_list_paginate(
```
Comment (clokep) [inline synapse/storage/database.py:1532]: This method actually isn't used, only `simple_select_list_paginate_txn` is used -- do we think we should just get rid of it?

**Comment by clokep** [comment]
See #8167 for the CI failures.

File: synapse/storage/database.py (hunk 1 above)
Comment (richvdh) [reply synapse/storage/database.py:1556]: yes!

File: synapse/storage/database.py [hunk 2]
```diff
@@ -605,7 +605,9 @@ def cursor_to_dict(cursor: Cursor) -> List[Dict[str, Any]]:
         results = [dict(zip(col_headers, row)) for row in cursor]
         return results
 
-    def execute(self, desc: str, decoder: Callable, query: str, *args: Any):
+    async def execute(
+        self, desc: str, decoder: "Optional[Callable[..., R]]", query: str, *args: Any
```
Comment (richvdh) [inline synapse/storage/database.py:609]:
```suggestion
        self, desc: str, decoder: "Optional[Callable[synapse.storage.types.Cursor, R]]", query: str, *args: Any
```

File: synapse/storage/database.py (hunk 1 above)
Comment (richvdh) [reply synapse/storage/database.py:1556]: (this was part of the disastrous #1784 which we have been gradually reverting for the last 3.5 years...)

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**